### PR TITLE
Feature/cdm line item

### PR DESCRIPTION
**Pull request by fivetran-catfritz**
## PR Overview
**This PR will address the following Issue/Feature:**
- Quarterly project

**This PR will result in the following new package version:**

- v0.3.0 for new model

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

> ## Features
> - Introduced the new `recharge__line_item_enhanced` model. This model includes a line item enriched with invoice, subscription, payment, and refund information. This model has been built with the intention of retaining a common line item schema across all other Fivetran billing data models.
> 
> ## 🚨 Breaking Changes 🚨
> - In the dbt_recharge_source v0.3.0 release, the following columns were added to model `stg_recharge__address`:
>   - country
>   - payment_method_id
>   - ❗Note: If you have already added any of these fields as passthrough columns to the `recharge__address_passthrough_columns` var, you will need to remove or alias these fields from the var to avoid duplicate column errors. Refer to the [source package's release notes](https://github.com/fivetran/dbt_recharge_source/releases/tag/v0.3.0) for more details.
> 
> ## Additional source package updates ([release notes](https://github.com/fivetran/dbt_recharge_source/releases/tag/v0.3.0))
> - Added staging model `stg_recharge__checkout`. See [this doc](https://fivetran.github.io/dbt_recharge/#!/model/model.recharge_source.stg_recharge__checkout) for the fields added and their definitions.
>   - This model is disabled by default but can be enabled by setting variable `recharge__checkout_enabled` to true in your `dbt_project.yml` file. See the [Enable/disable models and sources section](https://github.com/fivetran/dbt_recharge/blob/main/README.md#step-4-enable-disable-models-and-sources) of the README for more information.
>   - This model can also be passed additional columns beyond the predefined columns by using the variable `recharge__checkout_passthrough_columns`. See the [Passing Through Additional Columns](https://github.com/fivetran/dbt_recharge/blob/main/README.md#passing-through-additional-columns) section of the README for more information on how to set this variable.
> 
> - Added the following columns to model `stg_recharge__customer`. See [this doc](https://fivetran.github.io/dbt_recharge/#!/model/model.recharge_source.stg_recharge__customer) for field definitions.
>   - `billing_first_name`
>   - `billing_last_name`
>   - `billing_company`
>   - `billing_city`
>   - `billing_country`

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [ ] ~dbt run (if incremental models are present) && dbt test~

Before marking this PR as "ready for review" the following have been applied:
- [ ] ~The appropriate issue has been linked, tagged, and properly assigned~
- [ ] All necessary documentation and version upgrades have been applied
- [ ] docs were regenerated (unless this PR does not include any code or yml updates)
- [x] BuildKite integration tests are passing
- [x] Detailed validation steps have been provided below

### Detailed Validation
Please share any and all of your validation steps:

- See internal ticket for schema sample.

### If you had to summarize this PR in an emoji, which would it be?
 
:dancer:

### CODE REVIEW

File: models/common data models/recharge__line_item_enhanced.sql
Comment (fivetran-catfritz) [inline models/common data models/recharge__line_item_enhanced.sql:69]: I considered just making this line `'charge' as transaction_type` since everything in this table is a charge and would look more like the other packages, but I thought this was more useful for Recharge specifically. For example, a charge might be generally recurring, but it might have a onetime item that a customer added on. Open to

File: models/common data models/recharge__line_item_enhanced.sql
Comment (fivetran-joemarkiewicz) [inline models/common data models/recharge__line_item_enhanced.sql:46]: Very small request, but so far the datatypes for these two fields in the other models have been strings. In this platform they are integers. Would you be able to update the datatype for these to be strings to ensure uniformity of the model datatypes across platforms.

File: models/common data models/recharge__line_item_enhanced.sql
Comment (fivetran-joemarkiewicz) [inline models/common data models/recharge__line_item_enhanced.sql:67]: Not a required suggestion, but wanted to share a more consolidated way to achieve these results. Please note I have not tested this compilation.
```suggestion
        -- currency is in the charges api but not the fivetran schema, so relying on checkouts for now. this only has 20% utilization though so we will want to switch if they add it.
        cast({{ 'checkouts.currency' if var('recharge__using_checkout', false) else 'null' }} as {{ dbt.type_string() }}) as currency,
```

File: models/common data models/recharge__line_item_enhanced.sql
Comment (fivetran-joemarkiewicz) [reply models/common data models/recharge__line_item_enhanced.sql:69]: I agree with your decision to have this be `purchase_item_type`. This fits better for the platform as whole and tells a more complete story.

File: CHANGELOG.md
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:4]: Be sure to include any breaking changes from the source and any new variables from the source in these CHANGELOG notes.

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-catfritz great work on this PR! I just a have a few requested changes before this is ready for full approval! Thanks again 😃

**Review by fivetran-catfritz — COMMENTED** [review]
@fivetran-joemarkiewicz Thanks for the comments! This is ready for re-review. I added as much documentation as I could minus the field definitions we will add later.

File: CHANGELOG.md
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:8]: Can you link to the source release here

File: README.md
Comment (fivetran-joemarkiewicz) [inline README.md:75]: Same comment from source, the default version of this variable in the code is `false`. Is this expected, or should it be `true` by default?

File: models/common_data_models/docs.md
Comment (fivetran-joemarkiewicz) [inline models/common_data_models/docs.md:2]: Would you actually be able to update these to match [the following descriptions](https://github.com/fivetran-streamlit/standardized_billing_model/blob/fa8af19e0c720792af04c06ca119acc522d2b5fd/pages/billing_schema.py#L31-L64).

File: models/common_data_models/recharge__common_data_models.yml
Comment (fivetran-joemarkiewicz) [inline models/common_data_models/recharge__common_data_models.yml:10]: Be sure to include the `source_relation` field here. I missed this originally but you will need to include that field in this test and also in all the joins within the new model.

File: models/common_data_models/docs.md
Comment (fivetran-joemarkiewicz) [reply models/common_data_models/docs.md:2]: In case you don't have access to the above repo:
```yml
schema = {
    "🔑 header_id": "ID of either the invoice or order.",
    "🔑 line_item_id": "ID of either the invoice line or order line item.",
    "line_item_index": "Numerical identifier of the line item within the header object.",
    "record_type": "Either 'header' or 'line_item' to differentiate if the record is originally from the line item table or was created to document information only available at the header level.",
    "created_at": "Date the line item entry was created.",
    "currency": "Determines the currency in which the transaction took place.",
    "header_status": "Indicates the status of the header. Eg. paid, voided, returned.",
    "product_id": "ID of the product associated with the line item.",
    "product_name": "Name of the product associated with the line item.",
    "product_type": "Type of the product (e.g., physical, digital).",
    "product_category": "Category to which the product belongs (e.g., electronics, clothing).",
    "quantity": "Quantity of the product in the line item.",
    "unit_amount": "Unit price of the product.",
    "discount_amount": "Amount of discount applied to the line item.",
    "tax_rate": "Tax rate applied to the line item.",
    "tax_amount": "Amount of tax applied to the line item.",
    "total_amount": "Total amount for the line item (including tax and discounts).",
    "payment_id": "ID of the payment associated with the line item.",
    "payment_method": "Method used for payment (e.g., credit card, PayPal).",
    "payment_at": "Date and time of the payment.",
    "fee_amount": "Any additional fees associated with the payment.",
    "refund_amount": "Amount refunded for the line item (if applicable).",
    "subscription_id": "ID of the subscription (if applicable).",
    "subscription_period_started_at": "Start date of the subscription period (if applicable).",
    "subscription_period_ended_at": "End date of the subscription period (if applicable).",
    "subscription_status": "Status of the subscription (if applicable).",
    "customer_id": "ID of the customer associated with the line item.",
    "customer_level": "Identifies if the customer is reported at the account or customer level.",
    "customer_name": "Name of the customer.",
    "customer_company": "Company name of the customer (if applicable).",
    "customer_email": "Email address of the customer.",
    "customer_city": "City of the customer's address.",
    "customer_country": "Country of the customer's address."
}
```

File: integration_tests/seeds/checkout_data.csv
Comment (fivetran-joemarkiewicz) [inline integration_tests/seeds/checkout_data.csv:1]: For some reason I keep getting this failure when trying to run this using the seed data and having the using checkout variable enabled as `true`. Any ideas?

image

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-catfritz thanks for these most recent changes! This is looking great and real close but I have just a few questions before approval.

File: CHANGELOG.md
Comment (fivetran-catfritz) [reply CHANGELOG.md]: Moved it to the front!

File: integration_tests/seeds/checkout_data.csv
Comment (fivetran-catfritz) [reply integration_tests/seeds/checkout_data.csv:1]: Ahh I needed to update the dbt_project.yml with the seed column type to match what I did in the source.

File: models/common_data_models/docs.md
Comment (fivetran-catfritz) [reply models/common_data_models/docs.md]: @fivetran-joemarkiewicz A couple questions:
1. Should I customize the definitions for the package or is the idea to have a single set of definitions? For example, `header_id` is defined here as an invoice or order, but in recharge it is neither. It is charge_id, and the order_id is a different table and not used in this model. 
2. I had already created a `docs.md` that has definitions with the same name, like `customer_id` for example and causes a dbt error. Maybe should we hard code the definitions into the yml for the cdms to avoid this issue in this package and others? 
3. Just a note--`billing type`, `transaction_type`, and `payment_method_id` are missing from the linked list.

File: models/common_data_models/recharge__common_data_models.yml
Comment (fivetran-catfritz) [reply models/common_data_models/recharge__common_data_models.yml:10]: Recharge is not yet a member of the union data club. 😄

**Review by fivetran-catfritz — COMMENTED** [review]
@fivetran-joemarkiewicz Thanks for the comments! I made the necessary updates and have one more question about the docs. Thanks!

File: models/common_data_models/docs.md
Comment (fivetran-joemarkiewicz) [reply models/common_data_models/docs.md]: @fivetran-catfritz let's let the definitions be similar to the standard definitions. However, to your point we should add some specifics for the platform (like the header id being charge_id). Yeah let's plan to have these be their own yml since I can see that conflicting with the base models.

Thanks for catching those, I can add them to the streamlit report.

File: models/common_data_models/recharge__common_data_models.yml
Comment (fivetran-joemarkiewicz) [reply models/common_data_models/recharge__common_data_models.yml:10]: It's not cool yet haha then nevermind, disregard my comment lol

**Review by fivetran-catfritz — COMMENTED** [review]
@fivetran-joemarkiewicz I have updated the doc yml and generated the dbt docs!

**Review by fivetran-joemarkiewicz — APPROVED** [review]
LGTM!

**Review by fivetran-catfritz — COMMENTED** [review]
Agree with selection for `customer_created_at` and `subscription_plan`.

**Review by fivetran-reneeli — APPROVED** [review]
looks good!

**Pull request merged**
